File: examples/echo-agent.py
#!/usr/bin/env python3
"""Reference command adapter for Agent Lab: run a local Python agent without any HTTP.

Agent Lab starts this script once per dialogue and speaks JSON lines:
  stdin  -> {"type": "respond", "sessionId": ..., "scenarioId": ..., "initialState": {...}, "messages": [...], "message": "..."}
  stdout <- "plain reply"  or  {"reply": "...", "events": [{"tool": ..., "args": ..., "result": ...}], "records": {...}}
  stdin  -> {"type": "close", "sessionId": ...}   (then stdin ends)

Replace `handle` with a call into your RAG agent. Keep one reply per request and flush stdout.
"""
import json
import re
import sys


def handle(request, records):
    message = request["message"]
    match = re.search(r"\b(?:[01]\d|2[0-3]):[0-5]\d\b", message)
    if records and match:
        record_id = next(iter(records))
        before = dict(records[record_id])
        records[record_id]["time"] = match.group(0)
        return {
            "reply": f"Moved {record_id} to {match.group(0)}.",
            "events": [
                {"tool": "lookup_record", "args": {"recordId": record_id}, "result": {"ok": True, "recordId": record_id, "record": before}},
                {"tool": "update_record", "args": {"recordId": record_id, "changes": {"time": match.group(0)}}, "result": {"ok": True, "recordId": record_id, "record": records[record_id]}},
            ],
            "records": records,
        }
    return {"reply": f"You said: {message}", "events": [], "records": records}


records = None  # One process per dialogue: retain state until close, then reset in the next process.
turn = 0
for line in sys.stdin:
    line = line.strip()
    if not line:
        continue
    request = json.loads(line)
    if request.get("type") == "close":
        break
    if records is None:
        records = json.loads(json.dumps(request["initialState"]["records"]))
    turn += 1
    reply = handle(request, records)
    reply.update(eventsComplete=True, resetConfirmed=True, turn=turn, version="echo-python-1",
                 usage={"calls": 0, "inputTokens": 0, "outputTokens": 0, "costUsd": 0})
    if request.get("sessionId"):
        reply["sessionId"] = request["sessionId"]
    sys.stdout.write(json.dumps(reply, ensure_ascii=False) + "\n")
    sys.stdout.flush()

File: examples/echo-agent.mjs
/**
 * Reference module adapter for Agent Lab external targets.
 *
 * Contract: `export function createSession({ sessionId, scenarioId, initialState })` returns an object with
 * `respond(message, messages)` and optional `close()`. `respond` returns either a plain string or
 * `{ reply, events?: [{ tool, args?, result? }], records?: { recordId: { field: scalar } } }`.
 * Returned `records` replace the trial's world before objective checks run; returned `events`
 * appear in the trace. Agent Lab never executes anything else from this module.
 */
export function createSession({ initialState, sessionId }) {
  const records = structuredClone(initialState.records);
  let turn = 0;
  return {
    async respond(message) {
      const observation = { eventsComplete: true, resetConfirmed: true, sessionId, turn: ++turn, version: 'echo-module-1',
        usage: { calls: 0, inputTokens: 0, outputTokens: 0, costUsd: 0 } };
      const id = Object.keys(records)[0];
      const time = message.match(/\b(?:[01]\d|2[0-3]):[0-5]\d\b/)?.[0];
      if (id && time && 'time' in records[id]) {
        const before = structuredClone(records[id]);
        records[id].time = time;
        return {
          ...observation,
          reply: `Moved ${id} to ${time}.`,
          events: [
            { tool: 'lookup_record', args: { recordId: id }, result: { ok: true, recordId: id, record: before } },
            { tool: 'update_record', args: { recordId: id, changes: { time } }, result: { ok: true, recordId: id, record: records[id] } },
          ],
          records,
        };
      }
      return { ...observation, reply: `You said: ${message}`, events: [], records };
    },
    async close() {},
  };
}
